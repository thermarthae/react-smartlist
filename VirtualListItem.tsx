import React, {
	ElementType,
	createElement,
} from 'react';
import {
	CallbackNode,
	unstable_scheduleCallback as scheduleCallback,
	unstable_cancelCallback as cancelCallback,
	unstable_UserBlockingPriority as UserBlockingPriority,
	unstable_LowPriority as LowPriority,
} from 'scheduler';

import { TEntry } from './VirtualList';

export type TChildrenProps<Item extends object = {}, Ref extends HTMLElement = HTMLElement> = {
	ref: React.Ref<Ref>;
	data: Item;
	rootElProps: {
		'data-index': number;
		'data-measured': boolean;
		style: {
			opacity?: number;
			willChange?: 'transform';
			position: 'absolute';
			width: '100%';
			transform: string;
			contain: 'content',
		};
	};
};

type TProps<I extends object = {}, C extends ElementType = ElementType> = {
	component: C;
	itemData: I;
	itemIndex: number;
	nailPoint: number;
	itWasMeasured: boolean;
	onMeasure: (item: TEntry<I>) => void;
	sharedProps?: Omit<React.ComponentPropsWithoutRef<C>, 'data' | 'rootElProps' | 'children'>;
};

class VirtualListItem<I extends object, C extends ElementType> extends React.PureComponent<TProps<I, C>> {
	private itemElRef = React.createRef<HTMLElement>();

	private resizeObserver: ResizeObserver | null = null;

	private scheduledObserver: CallbackNode | null = null;

	public componentDidMount() {
		this.attachResizeObserver();
	}

	public componentWillUnmount() {
		this.detachResizeObserver();
	}

	private attachResizeObserver = () => {
		const priority = !this.props.itWasMeasured ? UserBlockingPriority : LowPriority;
		this.scheduledObserver = scheduleCallback(priority, () => {
			if (!this.itemElRef.current) return;
			this.resizeObserver = new ResizeObserver(this.measureHeight);
			this.resizeObserver.observe(this.itemElRef.current);
		});
	};

	private detachResizeObserver = () => {
		if (this.scheduledObserver) cancelCallback(this.scheduledObserver);
		this.resizeObserver?.disconnect();
		this.resizeObserver = null;
	};

	private measureHeight: ResizeObserverCallback = ([entry]) => {
		const { onMeasure, itemIndex, itemData } = this.props;

		const height = entry.borderBoxSize[0].blockSize;
		if (height === 0) {
			return;
		}
		onMeasure({
			index: itemIndex,
			data: itemData,
			height,
		});
	};

	public render() {
		const {
			component,
			itemData,
			sharedProps,
			itWasMeasured,
			nailPoint,
			itemIndex,
		} = this.props;

		return createElement(
			component,
			{
				...(sharedProps || {}),
				data: itemData,
				ref: this.itemElRef,
				rootElProps: {
					'data-index': itemIndex,
					'data-measured': itWasMeasured,
					style: {
						opacity: itWasMeasured ? undefined : 0.5,
						willChange: itWasMeasured ? undefined : 'transform',
						position: 'absolute',
						width: '100%',
						transform: `translateY(${nailPoint}px)`,
						contain: 'content',
					},
				},
			} as TChildrenProps<I>,
		);
	}
}

export default VirtualListItem;
