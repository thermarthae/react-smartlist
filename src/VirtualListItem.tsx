import {
	ElementType,
	memo,
	useCallback,
	useEffect,
	useRef,
} from 'react';
import {
	unstable_cancelCallback as cancelCallback,
	unstable_LowPriority as LowPriority,
	unstable_scheduleCallback as scheduleCallback,
	unstable_UserBlockingPriority as UserBlockingPriority,
} from 'scheduler';
import { shallowEqualObjects } from 'shallow-equal';

export type TSharedProps<P> = Omit<P, keyof TChildrenProps | 'children'>;
export type TChildrenProps<Data extends object = object, El extends HTMLElement = HTMLElement> = {
	data: Data;
	isAlreadyMeasured: boolean;
	rootElProps: {
		ref: React.Ref<El>;
		'data-index': number;
		'data-measured': boolean;
		style: {
			position: 'absolute';
			width: '100%';
			transform: string;
			contain: 'content';
		};
	};
};

export type TProps<Data extends object = object, C extends ElementType = ElementType> = {
	component: C;
	itemData: Data;
	itemIndex: number;
	nailPoint: number;
	sharedProps?: TSharedProps<React.ComponentPropsWithoutRef<C>>;
	isAlreadyMeasured: boolean;
	isMeasurmentDisabled?: boolean;
	onMeasure: {
		/**  React memo helper to check whether `onMeasure` actually differs */
		key?: string;
		(height: number): void;
	};
};

function VirtualListItem<Data extends object, C extends ElementType>({
	component,
	itemData,
	itemIndex,
	nailPoint,
	sharedProps,
	isAlreadyMeasured,
	isMeasurmentDisabled,
	onMeasure,
}: TProps<Data, C>) {
	const ref = useRef<HTMLElement>(null);
	const observer = useRef<ResizeObserver | null>(null);

	const onMeasureRef = useRef(onMeasure); // TODO: mayby useEffectEvent?
	useEffect(() => { onMeasureRef.current = onMeasure; }, [onMeasure]);

	const handleResize = useCallback<ResizeObserverCallback>(([entry]) => {
		const height = entry.borderBoxSize[0].blockSize;
		if (height === 0) return;

		onMeasureRef.current(height);
	}, []);

	useEffect(() => {
		if (!isMeasurmentDisabled) return () => {
			observer.current?.disconnect();
			observer.current = null;
		};
	}, [isMeasurmentDisabled, handleResize]);

	useEffect(() => {
		if (isMeasurmentDisabled || observer.current) return;

		// Use lower priority for already cached items
		const priority = isAlreadyMeasured ? LowPriority : UserBlockingPriority;
		const node = scheduleCallback(priority, () => {
			if (!ref.current) return;
			observer.current = new ResizeObserver(handleResize);
			observer.current.observe(ref.current);
		});

		return () => cancelCallback(node);
	}, [isMeasurmentDisabled, isAlreadyMeasured, handleResize]);

	const Child = component as React.ComponentType<TChildrenProps<Data>>;
	return (
		<Child
			{...{
				...sharedProps,
				data: itemData,
				isAlreadyMeasured,
				rootElProps: {
					ref,
					'data-index': itemIndex,
					'data-measured': isAlreadyMeasured,
					style: {
						position: 'absolute',
						width: '100%',
						transform: `translateY(${nailPoint}px)`,
						contain: 'content',
					},
				},
			}}
		/>
	);
}

export default memo(VirtualListItem, (prev, next) => {
	const { itemData: data, sharedProps: SP, onMeasure, ...prevRest } = prev;
	const { itemData: nextData, sharedProps: nextSP, onMeasure: nextOnMeasure, ...nextRest } = next;

	if (
		onMeasure.key !== nextOnMeasure.key
		|| !shallowEqualObjects(prevRest, nextRest)
		|| !shallowEqualObjects(data, nextData)
		|| !shallowEqualObjects(SP, nextSP)
	) return false;

	return true;
});
