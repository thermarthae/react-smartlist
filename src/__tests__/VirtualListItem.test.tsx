import {
	act,
	render,
} from '@testing-library/react';
import scheduler, {
	CallbackNode,
	unstable_hasPendingWork as hasPendingWork,
	unstable_LowPriority as LowPriority,
	unstable_scheduleCallback as scheduleCallback,
	unstable_UserBlockingPriority as UserBlockingPriority,
} from 'scheduler/unstable_mock';
import {
	describe,
	expect,
	it,
	vi,
} from 'vitest';

import type { TEntry } from '../VirtualList.tsx';
import VirtualListItem, { TChildrenProps, TProps as TVirtualListItemProps } from '../VirtualListItem.tsx';

type TItem = { id: number };

vi.useFakeTimers();

describe('VirtualListItem', () => {
	const getLastScheduledNode = () => vi.mocked(scheduleCallback).mock
		.results.at(-1)?.value as CallbackNode | undefined;

	const onMeasureFn = vi.fn<(item: TEntry<TItem>) => void>();
	const ItemComponent = vi.fn(({
		rootElProps,
		innerRef,
		height,
		...rest
	}: TChildrenProps<TItem, HTMLDivElement> & { height?: number }) => (
		<div
			{...rootElProps}
			style={{
				...rootElProps.style,
				height,
			}}
			ref={innerRef}
			children={JSON.stringify(rest)}
		/>
	));
	const defaultProps: TVirtualListItemProps<TItem, typeof ItemComponent> = {
		component: ItemComponent,
		itWasMeasured: false,
		itemID: 0,
		itemData: { id: 0 },
		itemIndex: 0,
		nailPoint: 0,
		onMeasure: onMeasureFn,
	};

	const triggerMeasurement = () => act(() => {
		vi.runAllTimers();
		scheduler.unstable_flushAll();
	});

	it('should render', () => {
		const { container } = render(<VirtualListItem {...defaultProps} />);

		expect(ItemComponent).toHaveBeenCalled();
		expect(container.firstChild).toMatchSnapshot();
	});

	it('should attach observers with correct priorities', () => {
		// first render has an UserBlockingPriority:
		render(<VirtualListItem {...defaultProps} />);
		expect(getLastScheduledNode()?.priorityLevel).toBe(UserBlockingPriority);

		triggerMeasurement();

		// everything except the first render has LowPriority:
		render(<VirtualListItem {...defaultProps} itWasMeasured />);
		expect(getLastScheduledNode()?.priorityLevel).toBe(LowPriority);
	});

	it('should not attach observers when measurment is disabled', () => {
		render(<VirtualListItem {...defaultProps} isMeasurmentDisabled />);
		expect(hasPendingWork()).toBeFalsy();
	});

	it('should attach observers when measurment is no longer disabled', () => {
		const { rerender } = render(<VirtualListItem {...defaultProps} isMeasurmentDisabled />);
		expect(hasPendingWork()).toBeFalsy();

		rerender(<VirtualListItem {...defaultProps} isMeasurmentDisabled={false} />);
		expect(hasPendingWork()).toBeTruthy();
	});

	it('should abort an observer attachment at premature unmount', () => {
		const { unmount } = render(<VirtualListItem {...defaultProps} />);

		const node = getLastScheduledNode()!;
		expect(node.callback).toBeDefined();

		unmount();
		expect(node.callback).toBeNull();
	});

	it('should not trigger the onMeasure event when height is equal to 0', () => {
		render(<VirtualListItem {...defaultProps} sharedProps={{ height: 0 }} />);

		// give a scheduler some time to attach the listener
		triggerMeasurement();

		expect(onMeasureFn).not.toHaveBeenCalled();
	});

	it('should trigger the onMeasure event', () => {
		render(<VirtualListItem {...defaultProps} sharedProps={{ height: 1 }} />);

		// give a scheduler some time to attach the listener
		triggerMeasurement();

		expect(onMeasureFn).toHaveReturnedTimes(1);
		expect(onMeasureFn).toHaveBeenCalledWith({
			id: defaultProps.itemID,
			index: defaultProps.itemIndex,
			data: defaultProps.itemData,
			height: 1,
		});
	});

	it('should not rerender when it is unnecessary', () => {
		let prevProps: TVirtualListItemProps<TItem> = {
			...defaultProps,
			sharedProps: {},
		};
		const sCU = vi.spyOn(VirtualListItem.prototype, 'shouldComponentUpdate');
		const { rerender } = render(<VirtualListItem {...prevProps} />);

		const testProps = (newProps: Partial<typeof prevProps>, shouldRerender: boolean) => {
			sCU.mockClear();
			prevProps = { ...prevProps, ...newProps };
			rerender(<VirtualListItem {...prevProps} />);

			expect(sCU).toHaveNthReturnedWith(1, shouldRerender);
		};

		testProps({}, false);
		testProps({ itemData: { ...prevProps.itemData } }, false);
		testProps({ itemData: { id: 52 } }, true);
		testProps({ itemIndex: 43 }, true);
		testProps({ nailPoint: 91 }, true);
		testProps({ itWasMeasured: !prevProps.itWasMeasured }, true);
		testProps({ onMeasure: () => 0 }, true);
		testProps({ sharedProps: { ...prevProps.sharedProps } }, false);
		testProps({ sharedProps: { x: 2 } }, true);
	});
});
