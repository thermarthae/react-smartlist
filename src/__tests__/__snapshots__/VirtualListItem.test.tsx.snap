// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VirtualListItem > should render 1`] = `
<div
  data-index="0"
  data-measured="false"
  style="position: absolute; width: 100%; transform: translateY(0px); contain: content;"
>
  {"data":{"id":0},"itWasMeasured":false}
</div>
`;
